### PR TITLE
diffstarpop_halpha

### PR DESCRIPTION
**Pull request by zaidikumail**
Get halpha for diffstarpop

### CODE REVIEW

**Review by zaidikumail — COMMENTED** [review]
This does not fix the problem of having large values for loss when they should be zero.

**Comment by zaidikumail** [comment]
Now onwards, making small changes locally and testing before committing to avoid an unnecessarily long list of commits

**Comment by zaidikumail** [comment]
@aphearin - This PR can now be merged into main. The following has been implemented:

- diffstarpop default parameters can now be successfully translated into composite (smooth main sequence+quiescent) H-alpha Luminosity Function. Dust and Burstiness remain to be incorporated.
- diffstarpop default parameters can be successfully recovered using Adam gradient descent, starting from an initially perturbed subset of diffstarpop defaults.
- as an aside: previously implemented `bimodal_sfh_opt.py` in #1 now correctly uses the composite LF to calculate MSE instead of the SF and Q ones.

File: diffhtwo/experimental/diffstarpop_halpha_opt.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+from jax.flatten_util import ravel_pytree
+from diffstarpop.defaults import DEFAULT_DIFFSTARPOP_PARAMS
+from jax import lax
+from functools import partial
+from .diffstarpop_halpha import diffstarpop_halpha_kern as dpop_halpha
+from .diffstarpop_halpha import (
+    diffstarpop_halpha_lf_weighted as dpop_halpha_lf_weighted,
+)
+
+theta0, unravel_fn = ravel_pytree(DEFAULT_DIFFSTARPOP_PARAMS)
```
Comment (aphearin) [inline diffhtwo/experimental/diffstarpop_halpha_opt.py]: Wow - what an exceptionally useful function! I had no idea this existed.

File: diffhtwo/experimental/diffstarpop_halpha_opt.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+
+def make_subspace_loss(unravel_fn, theta_default_flat, idx):
+    """
+    Build a loss that optimizes ONLY the parameters at flat indices `idx`.
+    - unravel_fn: from ravel_pytree(template)
+    - theta_default_flat: 1D base vector (others stay fixed to these values)
+    - idx: 1D array/list of flat indices to vary (static for the compiled fn)
+
+    Notes: The only thing you should not do is use namedtuple._replace() / ._asdict()
```
Comment (aphearin) [inline diffhtwo/experimental/diffstarpop_halpha_opt.py:43]: The `jax.flatten_utils.ravel_pytree` function trick looks so handy that I'm going to try doing things this way going forwards. I have used this pattern within gradient descents before, but I did not know that `namedtuple._replace() / ._asdict()` created a lot of slowdown for the jit!

File: diffhtwo/experimental/diffstarpop_halpha_opt.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+from diffstarpop.defaults import DEFAULT_DIFFSTARPOP_PARAMS
+from jax import lax
+from functools import partial
+from .diffstarpop_halpha import diffstarpop_halpha_kern as dpop_halpha
+from .diffstarpop_halpha import (
+    diffstarpop_halpha_lf_weighted as dpop_halpha_lf_weighted,
+)
+
+theta0, unravel_fn = ravel_pytree(DEFAULT_DIFFSTARPOP_PARAMS)
+idx = jnp.array([1, 8, 16])
```
Comment (aphearin) [inline diffhtwo/experimental/diffstarpop_halpha_opt.py]: Minor style comment: when you are defining a global constant like this, try to use `ALL_CAPS` for the variable. This is a PEP8 style convention that helps readability.

File: diffhtwo/experimental/diffstarpop_halpha_opt.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+
+@jjit
+def _mse(halpha_lf_weighted_composite_true, halpha_lf_weighted_composite_pred):
+    diff = halpha_lf_weighted_composite_true - halpha_lf_weighted_composite_pred
+    return jnp.mean(jnp.square(diff))
+
+
+def make_subspace_loss(unravel_fn, theta_default_flat, idx):
```
Comment (aphearin) [inline diffhtwo/experimental/diffstarpop_halpha_opt.py]: Clever use of a closure pattern here.

File: diffhtwo/experimental/halpha_luminosity.py [hunk 5]
```diff
@@ -52,20 +52,61 @@ def get_L_halpha(
 )
 
 
+def safe_log10(x):
```
Comment (aphearin) [inline diffhtwo/experimental/halpha_luminosity.py:56]: Nice. Little tricks like this are a big part of becoming an expert JAX practitioner. This little kernel should be jitted though.

File: diffhtwo/experimental/tests/test_diffstarpop_halpha.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            0.35032972,
+            0.22578957,
+            0.14377021,
+            0.09140591,
+        ]
+    )
+    noise_std = 2
+    ssp_halpha_luminosity = [
+        np.clip(
+            arr + np.random.normal(0, noise_std, size=arr.shape), a_min=0, a_max=None
```
Comment (aphearin) [inline diffhtwo/experimental/tests/test_diffstarpop_halpha.py:79]: In unit tests where there is stochasticity, I try to use `jax.random` rather than `np.random`, because then the test has fully deterministic behavior. Otherwise, a unit test can sometimes pass, and other times fail, depending the random seed used in the testing, which can create confusion. But this is a very minor thing and not worth changing here.

File: diffhtwo/experimental/tests/test_diffstarpop_halpha.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    t_table = np.linspace(T_TABLE_MIN, 10**lgt0, 100)
+
+    mzr_params = umzr.DEFAULT_MZR_PARAMS
+
+    spspop_params = spspu.DEFAULT_SPSPOP_PARAMS
+    # scatter_params = DEFAULT_SCATTER_PARAMS
+    # ssp_err_pop_params = ssp_err_model.DEFAULT_SSPERR_PARAMS
+
+    args = (
```
Comment (aphearin) [inline diffhtwo/experimental/tests/test_diffstarpop_halpha.py:116]: Watch out for using the same `jax.random.key`, which here is being used for diffstarpop, and a few lines above it is used for diffmahpop. Once you use a random key once, you should not reuse it.

The pattern JAX recommends goes something like this.

```
ran_key = jax.random.key(0)
ran_key, diffmahpop_key = jran.split(ran_key, 2)
result = diffmahpop_func(diffmahpop_key, *args)
[...]
ran_key, diffstarpop_key = jran.split(ran_key, 2)
result = diffstarpop_func(diffstarpop_key, *args)
[...]
```
Doing things ^this way, you have a running "primary" key, `ran_key`, that you keep splitting each time you do some new stochastic thing, which helps your code retain deterministic behavior when passed the same seed, even as you append new functionality to it.

An alternative pattern that is equivalent to above, and is handy when you know in advance that you'll need multiple keys:
```
ran_key = jax.random.key(0)
ran_key, diffmahpop_key, diffstarpop_key = jran.split(ran_key, 3)
result = diffmahpop_func(diffmahpop_key, *args)
[...]
result = diffstarpop_func(diffstarpop_key, *args)
[...]
```

File: diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+
+def _jitter_value(v, s, rng):
+    if isinstance(v, Real):
+        eps = s * (abs(v) if v != 0 else 1.0)
+        return v + rng.normal(0.0, eps)
+    return v
+
+
+def perturb_namedtuple(nt, scale=0.01, rng=None, bounds=None, per_field_scale=None):
```
Comment (aphearin) [inline diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py]: I guess this function came from the internet? Or maybe a bot? I think I see what it's doing. But now that you have a handy one-liner to create a flattened array from a nested namedtuple, is this function actually necessary?

File: diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+
+def perturb_diffstarpop(params, scale=0.01, rng=None):
+    """
+    Convenience wrapper for your DiffstarPopParams-like namedtuple,
+    with sensible default bounds for probability-like params and stds.
+    """
+    # Heuristic bounds for common fields (edit to taste)
+    bnds = {}
+    # Anything that looks like a fraction or ylo/yhi: clamp to [0, 1]
```
Comment (aphearin) [inline diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py]: Ok, this is important: DiffstarPop actually has bounds defined for each parameter. The same is true for all diffsky model components. And the are not always square, and certainly not always between 0 and 1. We should really be running these gradient descents on the _unbounded_ parameters rather than the bounded ones.

File: .gitignore [hunk 10]
```diff
@@ -1,5 +1,5 @@
 *.hdf5
-
+*.ipynb
```
Comment (aphearin) [inline .gitignore:2]: Generally a good idea to exclude notebook files from a repo like you are doing here. In case you want to include a notebook in future, just be sure to only add _unrendered_ notebooks, i.e., first do "restart kernel and clear all cells" and resave the notebook before doing `git add`. An unrendered notebook only occupies a small file size, where as a rendered one be pretty large, so this helps minimize the size of a source-code repo.

**Review by aphearin — CHANGES_REQUESTED** [review]
This is a really nice PR! Most of my comments are minor and many don't necessitate any change. The only significant one is that we should be using _unbounded_ diffstarpop parameters in our gradient descents, and using the bounding/unbounding functions to transform back and forth. That's quite important so let's not merge this PR until that change is made.

File: diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py [hunk 11]
```diff
@@ -165,7 +165,7 @@ def test_bimodal_sfh_opt():
         spspop_params,
     )
 
-    loss_hist, theta_best_fit = fit_diffstarpop(*fit_args, n_steps=1000, step_size=1e-2)
+    loss_hist, theta_best_fit = fit_diffstarpop(*fit_args, n_steps=2000, step_size=1e-3)
```
Comment (aphearin) [inline diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py]: In case this particular unit test takes an annoyingly long time to terminate (which looks like it may be the case, since you are using 2000 steps), you can always just decrease the tolerance required for agreement. We don't necessarily need every last science validation test to be included in our CI, only the ones that are most important for protecting our code against new bugs (and/or against regressions to previously-resolved bugs).

Another trick I sometimes use is to have _some_ tests only run on my own personal machine, and these tests get skipped when run up on the github servers. [Here is an example in DSPS](https://github.com/ArgonneCPAC/dsps/blob/2cbd7d1beb5e510400261d4b3f0180725d7e82d4/dsps/data_loaders/tests/test_load_filter_data.py#L26) where I only run the test if the `DSPS_DRN` environment variable exists on the machine, and otherwise it is skipped. You could use a similar trick if you wanted to avoid very-long-running unit tests on CI. The only downside is you have to remember to run `pytest` locally from time to time.

**Comment by zaidikumail** [comment]
@aphearin - Now, the gradient descents are performed in the unbounded param space and converted back to bounded params using a diffstarpop function when needed. Let's merge this PR into main now, if you agree. Next, I'll create a new PR to introduce burstiness!

As for the `test_diffstarpop_halpha_opt.py`, I have kept the `atol` high and `n_steps` low so that I can get a successful result sooner. Maybe later, we can skip this particular test if needed, as it still takes about 13 minutes for all the tests to run.

File: diffhtwo/experimental/diffstarpop_halpha_opt.py [hunk 12]
```diff
@@ -23,7 +23,7 @@
     diffstarpop_halpha_lf_weighted as dpop_halpha_lf_weighted,
 )
 
-theta0, unravel_fn = ravel_pytree(DEFAULT_DIFFSTARPOP_U_PARAMS)
+u_theta_default, u_unravel_fn = ravel_pytree(DEFAULT_DIFFSTARPOP_U_PARAMS)
```
Comment (aphearin) [inline diffhtwo/experimental/diffstarpop_halpha_opt.py:26]: Yes, thanks for being conscientious about using the `u_` convention. I promise this helps protect against confusion down the road!

**Comment by aphearin** [comment]
Nice job adapting quickly to the unbound space! Yes, this looks good to merge to me.

Ten minutes is indeed a really long time to wait for tests to run. We should probably find another workaround to be able to run the tests more quickly, while still retaining most of the validation of the test. Certainly fine by me to just kick this can down the road for now.

**Review by aphearin — APPROVED** [review]
LGTM!
